src/Middleware/RetryMiddleware.php

```diff
@@ -86,12 +95,27 @@ public function __invoke(Call $call, array $options)
         }
 
         return $nextHandler($call, $options)->then(null, function ($e) use ($call, $options) {
-            if (!$e instanceof ApiException) {
-                throw $e;
-            }
-
-            if (!in_array($e->getStatus(), $this->retrySettings->getRetryableCodes())) {
-                throw $e;
+            $customRetryFunc = $this->retrySettings->getRequestRetryFunction();
+            // If we have a custom retry function, and it returns false
+            // we throw the exception, otherwise we continue.
+            if (!is_null($customRetryFunc)) {
+                if (!$customRetryFunc($e, $options, $this->currentAttempts)) {
```

Comment:
The function signature of the custom retry function is not documented anywhere. It seems like it should be something like `function (Exception $e, RetrySettings $options, int $retryAttempt): bool` ? 

For guidance, take a look at the [exponential backoff](https://github.com/googleapis/google-cloud-php/blob/main/Core/src/ExponentialBackoff.php#L100)

Reply:
Thanks for the suggestion.